test/test_helper/common.bash

```diff
@@ -205,3 +206,34 @@ function wait_for_empty_mail_queue_in_container() {
     # shellcheck disable=SC2016
     repeat_in_container_until_success_or_timeout "${TIMEOUT}" "${CONTAINER_NAME}" bash -c '[[ $(mailq) == *"Mail queue is empty"* ]]'
 }
+
+# Common defaults appropriate for most tests, override vars in each test when necessary.
+# TODO: Check how many tests need write access. Consider using `docker create` + `docker cp` for easier cleanup.
+function init_with_defaults() {
+  # Ignore absolute dir path and file extension, only extract filename:
+  export TEST_NAME
+  TEST_NAME="$(basename "${BATS_TEST_FILENAME}" '.bats')"
+
+  export PRIVATE_CONFIG
+  PRIVATE_CONFIG="$(duplicate_config_for_container . "${TEST_NAME}")"
+  export TEST_FILES_VOLUME="${PWD}/test/test-files:/tmp/docker-mailserver-test:ro"
+  export TEST_CONFIG_VOLUME="${PRIVATE_CONFIG}:/tmp/docker-mailserver:ro"
+
+  export TEST_FQDN='mail.my-domain.com'
+}
+
+# Common docker run command that should satisfy most tests which only modify ENV.
+function common_container_setup() {
+  local TEST_ENV_FILE=$1
+
+  run docker run -d --name "${TEST_NAME}" \
+    --volume "${TEST_FILES_VOLUME}" \
+    --volume "${TEST_CONFIG_VOLUME}" \
+    --hostname "${TEST_FQDN}" \
+    --env-file "${TEST_ENV_FILE}" \
+    --tty \
+    "${NAME}"
+  assert_success
+
+  wait_for_finished_setup_in_container "${TEST_NAME}"
+}
```

Comment:
GitHub says here is no newline character at the end of this file. Intended?

Reply:
It does that to me too randomly. Very odd when it happens.

Reply:
Possibly an auto-formatting issue?

Reply:
Nope not intended, usually my editor adds them on save. I noticed similar was happening with a PR from @NorseGaud , although I'm not sure if I was overly fussed about it, we may have some existing files that are missing new line at end of file, is there an easy way to check all files for it?

Reply:
I know that I copy/pasted this portion to the file from the bats test that I originally had it in. It's not a Github thing as my local commit has it missing too.

Looks like I need to tell VSCode to make sure it adds new lines on save, or get it to load the `.editorconfig` file, which I recall being a hassle with VSCode to get working properly in the past :\

Do either of you have `.editorconfig` working with VSCode? Otherwise another option is to have a local git commit hook or github workflow on merge that ensures compliance with `.editorconfig`, actually I thought we had a lint for this...?

Reply:
Ah.. it seems that we don't lint our tests for some reason..?

https://github.com/docker-mailserver/docker-mailserver/blob/ddd97274e03252647e85ecbcf3cb8996b51f34e3/test/linting/.ecrc.json#L4-L26

Reply:
Originally eclint excludes and eclint itself was introduced by @georglauterbach [here](https://github.com/docker-mailserver/docker-mailserver/pull/1619/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R176):

**Makefile:**

```makefile
@ if /usr/bin/eclint -exclude "\.bats$$" | grep .; then\
  echo -e "\nError" ;\
  exit 1 ;\
else\
  echo -e '\nSuccess' ;\
fi
```

Then a later PR also by @georglauterbach [migrated the eclint with exclude to `lint.sh`](https://github.com/docker-mailserver/docker-mailserver/pull/1636/files#diff-04533026e36c69e3060191462221731a0c191379a2eb9fcc77775283004c9436R84):

```sh
local LINT=(eclint -exclude "(.*\.git.*|.*\.md$|\.bats$)")
```

Where we can see the exclude was updated, and once more updated again by @georglauterbach [in this PR](https://github.com/docker-mailserver/docker-mailserver/pull/1654/files#diff-04533026e36c69e3060191462221731a0c191379a2eb9fcc77775283004c9436R85):

```sh
local LINT=(eclint -exclude "(.*\.git.*|.*\.md$|\.bats$|\.cf$|\.conf$|\.init$)")
```

Next up @georglauterbach [introduced the `.ecrc.json` eclint config file, and the `LINT` var for excludes was renamed to `IGNORE`](https://github.com/docker-mailserver/docker-mailserver/pull/2000/files#diff-04533026e36c69e3060191462221731a0c191379a2eb9fcc77775283004c9436R54):

```sh
local IGNORE='.*\.git.*|.*\.md$|\.bats$|\.cf$|'
IGNORE+='\.conf$|\.init$|.*test/.*|.*tools/.*'
```

Then we moved linting tools into Docker images where [I advised extracting the `IGNORE` from `lint.sh` to move into `.ecrc.json`](https://github.com/docker-mailserver/docker-mailserver/pull/2001#issuecomment-854479492), [this commit from that PR](https://github.com/docker-mailserver/docker-mailserver/pull/2001/commits/8a5b284fe6c08d9d4e59716b913f056bc8ce0bef#diff-aacce0e557e390fb1d1c9fcfb577b0176aff26899d88681e2199b32a118ccbb9R7) applies that, with the excludes shown above in previous comment. `

And now we're up to date :stuck_out_tongue: 

In each case the added excludes seem to lack any discussion/reasoning? (_except the adoption of `.ecrc.json` PR that had [a linked issue with @NorseGaud mentioning that `tools` had to be added to excludes](https://github.com/docker-mailserver/docker-mailserver/issues/1997#issuecomment-846614128)_) I can't seem to find myself making any mention about the change from `.*test/.*` to `^test/` but assume it had the same outcome.

Reply:
@polarathene There is a dedicated extension for VS Code & editorconfig which works well for me: <https://marketplace.visualstudio.com/items?itemName=EditorConfig.EditorConfig>

---

As for the linting, I do not remember why exactly I removed the bats files, but I can test it locally again.

Reply:
> but I can test it locally again.

Sure but lets not enforce the linting on that just yet (assuming many lint issues are raised as a result).

Reply:
> Do either of you have `.editorconfig` working with VSCode?

I'm using the same extension @georglauterbach mentioned and never had an issue before 💁🏻 

Name: EditorConfig for VS Code
Id: editorconfig.editorconfig
Description: EditorConfig Support for Visual Studio Code
Version: 0.16.4
Publisher: EditorConfig
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=EditorConfig.EditorConfig



> or github workflow on merge that ensures compliance with `.editorconfig`, actually I thought we had a lint for this...

Should be covered by https://github.com/docker-mailserver/docker-mailserver/blob/ddd97274e03252647e85ecbcf3cb8996b51f34e3/.github/workflows/linting.yml#L27
We need to check if this lint ever worked and which config it is using ;)

Reply:
> and never had an issue before 💁🏻

I haven't tried in a while, but I recall it not working properly with VSCode when I last tried, while my previou editor Atom had no issue. I'm on Linux, if one of you are also, then it's probably an issue with my system. I'll try that extension again at some point :)

Reply:
MacOS 💻

Reply:
`Linux swift 5.11.0-34-generic #36-Ubuntu SMP Thu Aug 26 19:22:09 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux` :D